src/GitHub.Exports/Helpers/ThreadingHelper.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System.Threading.Tasks;
+using ThreadHelper = Microsoft.VisualStudio.Shell.ThreadHelper;
+using GitHub.Extensions;
+using System.Runtime.CompilerServices;
+using System;
+using static Microsoft.VisualStudio.Threading.JoinableTaskFactory;
+using static Microsoft.VisualStudio.Threading.AwaitExtensions;
+
+namespace GitHub.Helpers
+{
+    public interface IAwaitable
+    {
+        IAwaiter GetAwaiter();
+    }
+
+    public interface IAwaiter : INotifyCompletion
+    {
+        bool IsCompleted { get; }
+        void GetResult();
+    }
+
+    public static class ThreadingHelper
+    {
+        /// <summary>
+        /// Switch to the UI thread using ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync
+        /// Auto-disables switching when running in unit test mode
+        /// </summary>
+        /// <returns></returns>
+        public static IAwaitable SwitchToMainThreadAsync()
+        {
+            return Guard.InUnitTestRunner() ?
```

Comment:
Another way to implement this that wouldn't require the Awaitable plumbing would be:

```
        public static async Task SwitchToMainThreadAsync()
        {
            if (!Guard.InUnitTestRunner())
            {
                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
            }
        }
```

I'm not sure about the perf considerations, but assuming that thread switching using async is relatively expensive, it might not have much of an impact?